Don't create renderer for foreignElement when it has no parent

chirags27 opened the pull request:
#### e806076966607b8327822663338316299261960b
<pre>
Don&apos;t create renderer for foreignElement when it has no parent
<a href="https://bugs.webkit.org/show_bug.cgi?id=241778">https://bugs.webkit.org/show_bug.cgi?id=241778</a>
rdar://102916021

Reviewed by Nikolas Zimmermann.

If the body is replaced by foreignElement, it results in a DOM which
doesn&apos;t have anything to render or layout, but it crashes WebKit because
we trip over an assert because isPaintOffsetCacheEnabled is not disabled
(which happens only when there is a SVGRoot). This change makes it so
that we don&apos;t generate the useless renderer in the first place, which is
what happens with other SVGElements.

* LayoutTests/svg/foreignObject/foreign-object-without-svgroot-crash-expected.txt: Added.
* LayoutTests/svg/foreignObject/foreign-object-without-svgroot-crash.html: Added.
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::rendererIsNeeded):

Canonical link: <a href="https://commits.webkit.org/259685@main">https://commits.webkit.org/259685@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e3c33bf9bbab3e6d0340dc384ce91a2d77c2fb3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/105563 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/14647 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38454 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/114809 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174954 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/109463 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16067 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/5870 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/97852 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/114643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111318 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12218 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95200 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/39713 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94082 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26837 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81417 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7934 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28192 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8321 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/4781 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14057 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/47736 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6693 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/9953 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp:
```diff
@@ -54,8 +56,22 @@ SVGForeignObjectElement& LegacyRenderSVGForeignObject::foreignObjectElement() co
+            return true;
+        parent = parent->parent();
+    }
+    return false;
```
```suggestion
    return ancestorsOfType<LegacyRenderSVGRoot>(*this).first();
```

I don't think it's possible to have RenderSVGRoot and LegacyRenderSVGForeignObject in the same tree. Either everything is legacy (LBSE disabled) or non-legacy (LBSE enabled).

Checking for LegacyRenderSVGRoot should be sufficient.

Also you can use ancestorsOfType instead of this loop.

nt1m replied:
Although, I'm surprised there isn't a pre-existing function to get the SVG root, seems like something that would be done fairly often. Maybe Said or Nikolas know if there is one?

nikolaszimmermann replied:
There is the legacy SVGRenderSupport::findTeeeRootObject() method. However I’d use ancestorsOfType for new code…

nt1m replied:
I think we should use `SVGRenderSupport::findTreeRootObject()` in that case, I would consider `LegacyRenderSVGForeignObject` to be old code.  This fix does not affect LBSE.

nt1m commented on Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp:
```diff
@@ -54,8 +56,22 @@ SVGForeignObjectElement& LegacyRenderSVGForeignObject::foreignObjectElement() co
+
 void LegacyRenderSVGForeignObject::paint(PaintInfo& paintInfo, const LayoutPoint&)
 {
+    if (!this->isUnderSVGRoot())
```
```suggestion
    if (!isUnderSVGRoot())
```

nt1m commented on Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp:
```diff
@@ -120,6 +136,11 @@ RenderBox::LogicalExtentComputedValues LegacyRenderSVGForeignObject::computeLogi
 
 void LegacyRenderSVGForeignObject::layout()
 {
+    if (!this->isUnderSVGRoot()) {
```
```suggestion
    if (!isUnderSVGRoot()) {
```

nt1m commented on LayoutTests/fast/svg/foreign-object-without-svgroot-crash.html:
```diff
@@ -0,0 +1,13 @@
+function testFunction() {
+    if(window.testRunner)
+        testRunner.dumpAsText()
+    try { document.scrollingElement.replaceWith(foreignObjectID); } catch { }
```
```suggestion
    if (window.testRunner)
        testRunner.dumpAsText()
    document.documentElement.replaceWith(foreignObject);
```

nt1m commented on LayoutTests/fast/svg/foreign-object-without-svgroot-crash.html:
```diff
@@ -0,0 +1,13 @@
+</script>
+<body onload="testFunction()">
+<svg>
+<foreignObject id="foreignObjectID">
```
```suggestion
<foreignObject id="foreignObject">
```

nt1m commented on LayoutTests/fast/svg/foreign-object-without-svgroot-crash.html:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
+<script>
+function testFunction() {
```
Can we just name this `test()`?

nikolaszimmermann replied:
I‘d go for something descriptive: both test / testFunction are too generic. „reparentForeignObject“ / „replaceBodyWithForeignObject“?

Furthermore, I would move the test in a LayoutTests/svg/xxx subdirectory.

nikolaszimmermann commented on Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.h:
```diff
@@ -60,6 +60,7 @@ class LegacyRenderSVGForeignObject final : public RenderSVGBlock {
 
     const AffineTransform& localToParentTransform() const override;
     AffineTransform localTransform() const override { return m_localTransform; }
+    bool isUnderSVGRoot() const;
```
I wonder if we need to generalize this mechanism. What happens if you try to re-parent e.g. a <g> container to become the body element? All LegacyRenderSVG… classes contain the same assertion in layout().

Do we have other mechanisms in place that prevent to move SVG elements to form unexpected render trees? You need to duplicate your test for <svg> (should work), <g> (should still assert) and also for <use> which has a history of subtle bugs.

chirags27 replied:
> Do we have other mechanisms in place that prevent to move SVG elements to form unexpected render trees?
Based on Slack chat with a few engineers, I don't think there are things in place to do that. 

I'm not sure if we should fix all of those issues with the same patch - My intention is to have a targeted fix for the bug in question, but if we think there are other improvements that can be done, maybe those should be separate changes?

Let me try out some of the other elements to see if that increases scope by a lot.

chirags27 replied:
@nikolaszimmermann  I tried out various other cases like:
```
<!-- <svg>
    <circle id="foreignObject" cx="5" cy="5" r="4"/>
</svg> -->

<!-- <svg>
    <path id="foreignObject"
    d="M 10,30
           A 20,20 0,0,1 50,30
           A 20,20 0,0,1 90,30
           Q 90,60 50,90
           Q 10,60 10,30 z" />
</svg> -->

<!-- <svg>
    <ellipse id="foreignObject" cx="100" cy="50" rx="100" ry="50" />
</svg> -->

<!-- <svg>
    <circle id="foreignObject" cx="5" cy="5" r="4"/>
</svg> -->

<!-- <svg>
    <rect id="foreignObject" width="100" height="100"/>
</svg> -->
```
And none of those lead to a crash, mainly because the render tree in those cases just has a top level RenderView, so it doesn't get to the layout for SVG elements. Looks like we can proceed with this change for foreignObject.

nikolaszimmermann reviewed (changes requested):
Thanks for the patch. The general idea is fine, but I think it needs to be extended to all SVG renderers, see the inline comments.

nt1m commented on Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp:
```diff
@@ -54,8 +56,16 @@ SVGForeignObjectElement& LegacyRenderSVGForeignObject::foreignObjectElement() co
+
 void LegacyRenderSVGForeignObject::paint(PaintInfo& paintInfo, const LayoutPoint&)
 {
+    if (!isUnderSVGRoot())
```
can we just use `SVGRenderSupport::findTreeRootObject()` instead of introducing a new method that does the same thing?

nikolaszimmermann reviewed (approved):
r=me if EWS is happy too :-)

chirags27 committed:
Don't create renderer for foreignElement when it has no parent
https://bugs.webkit.org/show_bug.cgi?id=241778
rdar://102916021

Reviewed by Nikolas Zimmermann.

If the body is replaced by foreignElement, it results in a DOM which
doesn't have anything to render or layout, but it crashes WebKit because
we trip over an assert because isPaintOffsetCacheEnabled is not disabled
(which happens only when there is a SVGRoot). This change makes it so
that we don't generate the useless renderer in the first place, which is
what happens with other SVGElements.

* LayoutTests/svg/foreignObject/foreign-object-without-svgroot-crash-expected.txt: Added.
* LayoutTests/svg/foreignObject/foreign-object-without-svgroot-crash.html: Added.
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::rendererIsNeeded):

Canonical link: https://commits.webkit.org/259685@main
